### PR TITLE
Derive outcomes.exit.Exit from SystemExit instead of KeyboardInterrupt

### PR DESCRIPTION
**Pull request by blueyed**
This is required for properly getting out of pdb, where
KeyboardInterrupt is caught in py36 at least.

Ref: https://github.com/pytest-dev/pytest/issues/1865#issuecomment-242599949

/cc @The-Compiler

### CODE REVIEW

File: src/_pytest/runner.py [hunk 1]
```diff
@@ -216,6 +217,9 @@ def __init__(self, func, when, treat_keyboard_interrupt_as_exception=False):
             else:
                 self.stop = time()
                 raise
+        except Exit:
```
Comment (RonnyPfannschmidt) [inline src/_pytest/runner.py]: this change makes the api boundaries of CallInfo even more of a problem
i would like to propose reshaping callinfo a bit before adding that to it

Comment (blueyed) [reply src/_pytest/runner.py]: Well, I assume that might take some time?

This is an rather old issue (talked about in 2016 already).
Do you have a good suggestion for https://github.com/pytest-dev/pytest/pull/4280 then?

Comment (RonnyPfannschmidt) [reply src/_pytest/runner.py]: i got a pr for the change almost ready, will push this evening

Comment (blueyed) [reply src/_pytest/runner.py]: For referrence, I am assuming it is this one: https://github.com/pytest-dev/pytest/pull/4299

Comment (RonnyPfannschmidt) [reply src/_pytest/runner.py]: it is - i still have no idea why it is broken now ^^ its bound to be very subtle

File: testing/python/raises.py [hunk 2]
```diff
@@ -197,7 +197,9 @@ class ClassLooksIterableException(Exception):
             pass
 
         with pytest.raises(
-            Failed, match="DID NOT RAISE <class 'raises.ClassLooksIterableException'>"
+            # NOTE with xdist: "DID NOT RAISE <class 'raises.TestRaises.test_raises_exception_looks_iterable.<locals>.ClassLooksIterableException'>" ?!  # noqa: E501
+            Failed,
+            match=r"DID NOT RAISE <class 'raises(\..*)*ClassLooksIterableException'>",
```
Comment (blueyed) [inline testing/python/raises.py]: Why this change, only with xdist?

File: src/_pytest/runner.py [hunk 3]
```diff
@@ -190,10 +191,11 @@ def check_interactive_exception(call, report):
...
     hookname = "pytest_runtest_" + when
     ihook = getattr(item.ihook, hookname)
+    reraise = (Exit,)
+    if not item.config.getvalue("usepdb"):
+        reraise += (KeyboardInterrupt,)
     return CallInfo.from_call(
-        lambda: ihook(item=item, **kwds),
-        when=when,
-        reraise=KeyboardInterrupt if not item.config.getvalue("usepdb") else (),
+        lambda: ihook(item=item, **kwds), when=when, reraise=reraise
```
Comment (blueyed) [inline src/_pytest/runner.py]: Worth a `# fmt: off` / `# fmt: on` block maybe to silence black here?  (if this code stays like this)

File: testing/python/raises.py (hunk 2 above)
Comment (blueyed) [reply testing/python/raises.py]: Not related to this PR, but also with merge-master-into-features: https://travis-ci.org/pytest-dev/pytest/jobs/465815039#L570

File: src/_pytest/main.py [hunk 4]
```diff
@@ -205,7 +205,7 @@ def wrap_session(config, doit):
             raise
         except Failed:
             session.exitstatus = EXIT_TESTSFAILED
-        except KeyboardInterrupt:
+        except (KeyboardInterrupt, exit.Exception):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/main.py]: this implies that this is a breaking change for externals

Comment (blueyed) [reply src/_pytest/main.py]: What do you suggest?
Should we have a custom exception that is still derived from KeyboardInterrupt (if possible)?

Comment (RonnyPfannschmidt) [reply src/_pytest/main.py]: @blueyed its not clear to me off hand, i wonder if having multiple base classes is acceptable while preparing for a breaking release (i dont see a way to deal out warnings about this one tho)

Comment (blueyed) [reply src/_pytest/main.py]: Ok.
I've just noticed that also deriving from KeyboardInterrupt goes against what this tries to do: pdb not catching pytest's Exit.

I think it is OK for a feature release, and plugins should really catch/handle `Exit` in the first place already.

File: src/_pytest/runner.py (hunk 3 above)
Comment (nicoddemus) [reply src/_pytest/runner.py]: IMHO that's fine 😁

File: src/_pytest/main.py (hunk 4 above)
Comment (nicoddemus) [reply src/_pytest/main.py]: > I think it is OK for a feature release, and plugins should really catch/handle Exit in the first place already.

I think so too. I don't think this will bring problems to anyone TBH.

**Comment by nicoddemus** [comment]
Sorry for the delay @blueyed!
